File: src/test/java/org/shsts/tinactory/datagen/content/Veins.java
package org.shsts.tinactory.datagen.content;

import javax.annotation.ParametersAreNonnullByDefault;
import net.minecraft.MethodsReturnNonnullByDefault;
import org.shsts.tinactory.content.material.OreVariant;
import org.shsts.tinactory.datagen.content.builder.VeinBuilder;

import static org.shsts.tinactory.content.AllMaterials.BANDED_IRON;
import static org.shsts.tinactory.content.AllMaterials.BAUXITE;
import static org.shsts.tinactory.content.AllMaterials.BLUE_TOPAZ;
import static org.shsts.tinactory.content.AllMaterials.CASSITERITE;
import static org.shsts.tinactory.content.AllMaterials.CHALCOPYRITE;
import static org.shsts.tinactory.content.AllMaterials.CINNABAR;
import static org.shsts.tinactory.content.AllMaterials.COAL;
import static org.shsts.tinactory.content.AllMaterials.DIAMOND;
import static org.shsts.tinactory.content.AllMaterials.EMERALD;
import static org.shsts.tinactory.content.AllMaterials.GALENA;
import static org.shsts.tinactory.content.AllMaterials.GARNIERITE;
import static org.shsts.tinactory.content.AllMaterials.GOLD;
import static org.shsts.tinactory.content.AllMaterials.GRAPHITE;
import static org.shsts.tinactory.content.AllMaterials.HEAVY_OIL;
import static org.shsts.tinactory.content.AllMaterials.ILMENITE;
import static org.shsts.tinactory.content.AllMaterials.LIGHT_OIL;
import static org.shsts.tinactory.content.AllMaterials.LIMONITE;
import static org.shsts.tinactory.content.AllMaterials.MAGNETITE;
import static org.shsts.tinactory.content.AllMaterials.NATURAL_GAS;
import static org.shsts.tinactory.content.AllMaterials.PYRITE;
import static org.shsts.tinactory.content.AllMaterials.REDSTONE;
import static org.shsts.tinactory.content.AllMaterials.RUBY;
import static org.shsts.tinactory.content.AllMaterials.SAPPHIRE;
import static org.shsts.tinactory.content.AllMaterials.SILVER;
import static org.shsts.tinactory.content.AllMaterials.SPHALERITE;
import static org.shsts.tinactory.content.AllMaterials.TIN;
import static org.shsts.tinactory.content.AllMaterials.TOPAZ;

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public final class Veins {
    public static void init() {
        VEINS.vein("chalcopyrite", 0.4)
            .primitive()
            .ore(CHALCOPYRITE, 0.8)
            .ore(PYRITE, 0.6)
            .build()
            .vein("limonite", 0.6)
            .base()
            .ore(LIMONITE, 0.8)
            .ore(BANDED_IRON, 0.4)
            .ore(GARNIERITE, 0.2)
            .build()
            .vein("coal", 0.3)
            .ore(COAL, 1)
            .ore(COAL, 0.4)
            .build()
            .vein("cassiterite", 0.2)
            .ore(TIN, 1)
            .ore(CASSITERITE, 0.3)
            .ore(TIN, 0.1)
            .build()
            .vein("redstone", 0.1)
            .ore(REDSTONE, 1)
            .ore(RUBY, 0.3)
            .ore(CINNABAR, 0.1)
            .build()
            .vein("magnetite", 0.4)
            .ore(MAGNETITE, 1)
            .ore(GOLD, 0.2)
            .ore(MAGNETITE, 0.2)
            .build()
            .vein("sulfide", 0.4)
            .ore(GALENA, 0.5)
            .ore(SPHALERITE, 0.6)
            .ore(SILVER, 0.3)
            .build()
            .vein("graphite", 0.2)
            .ore(GRAPHITE, 1)
            .ore(GRAPHITE, 0.3)
            .ore(DIAMOND, 0.1)
            .build()
            .vein("bauxite", 0.5)
            .ore(BAUXITE, 0.8)
            .ore(ILMENITE, 0.6)
            .build()
            .vein("oil", 0.3)
            .ore(NATURAL_GAS, 0.3)
            .ore(LIGHT_OIL, 0.4)
            .ore(HEAVY_OIL, 0.7)
            .build()
            .vein("gem", 0.1)
            .ore(EMERALD, 0.4)
            .ore(SAPPHIRE, 0.7)
            .build()
            .vein("gold", 0.2)
            .variant(OreVariant.NETHERRACK)
            .ore(GOLD, 0.5)
            .ore(TOPAZ, 0.3)
            .ore(BLUE_TOPAZ, 0.3)
            .build();
    }

    private static class VeinFactory {
        private int rank = 0;

        public VeinBuilder<VeinFactory> vein(String id, double rate) {
            return VeinBuilder.factory(this, id, rank++, rate);
        }
    }

    private static final VeinFactory VEINS = new VeinFactory();
}

File: src/test/kotlin/org/shsts/tinactory/datagen/content/material/Veins.kt
package org.shsts.tinactory.datagen.content.material

import org.shsts.tinactory.core.material.OreVariant
import org.shsts.tinactory.datagen.content.builder.VeinBuilder

object Veins {
    fun init() {
        stone()
        deepslate()
        netherrack()
    }

    private fun stone() {
        vein("chalcopyrite", 0.4) {
            primitive = true
            ore("chalcopyrite", 0.8)
            ore("pyrite", 0.6)
        }
        vein("limonite", 0.6) {
            baseOre = true
            ore("limonite", 0.8)
            ore("banded_iron", 0.4)
            ore("garnierite", 0.2)
        }
        vein("coal", 0.3) {
            ore("coal", 1.0)
            ore("coal", 0.4)
        }
        vein("cassiterite", 0.2) {
            ore("tin", 1.0)
            ore("cassiterite", 0.3)
            ore("tin", 0.1)
        }
        vein("redstone", 0.1) {
            ore("redstone", 1.0)
            ore("ruby", 0.3)
            ore("cinnabar", 0.1)
        }
    }

    private fun deepslate() {
        vein("magnetite", 0.4) {
            ore("magnetite", 1.0)
            ore("gold", 0.2)
            ore("magnetite", 0.2)
        }
        vein("sulfide", 0.4) {
            ore("galena", 0.5)
            ore("sphalerite", 0.6)
            ore("silver", 0.3)
        }
        vein("graphite", 0.2) {
            ore("graphite", 1.0)
            ore("graphite", 0.3)
            ore("diamond", 0.1)
        }
        vein("bauxite", 0.5) {
            ore("bauxite", 0.8)
            ore("ilmenite", 0.6)
        }
        vein("oil", 0.3) {
            ore("natural_gas", 0.3)
            ore("light_oil", 0.4)
            ore("heavy_oil", 0.7)
        }
        vein("gem", 0.1) {
            ore("emerald", 0.4)
            ore("sapphire", 0.7)
        }
    }

    private fun netherrack() {
        vein("gold", 0.2) {
            variant(OreVariant.NETHERRACK)
            ore("gold", 0.5)
            ore("topaz", 0.3)
            ore("blue_topaz", 0.3)
        }
    }

    private var rank = 0

    private fun vein(id: String, rate: Double, block: VeinBuilder.() -> Unit) {
        VeinBuilder(id, rank++, rate).apply {
            block()
            build()
        }
    }
}
